File: runtime/scripts/prompt_builder.py
"""
Prompt Builder — Cache-aware system prompt assembly for CareSupport.

Produces structured content blocks (static-first, dynamic-last) and a proper
messages array, enabling Claude's prefix caching.

Block ordering:
  1. SOUL.md (identity, reasoning framework)        — never changes
  2. Routing + Capabilities + Skills                 — rarely changes
  3. Response format + channel guidance + enforcement — never changes
  4. Lessons (global + family)                       — changes weekly
  5. Member identity + member context                — changes per member
     ── CACHE BREAKPOINT ──  (~4,987 tokens, clears 4,096 minimum)
  6. Family context (filtered by role)               — intent-driven: full/partial/none
  7. Current datetime + member metadata              — changes every call

Intent-driven loading (Phase 1 of interaction/execution separation):
  CareRouter classifies intent → prompt builder loads only relevant family sections.
  EMERGENCY/ESCALATION get full context. GENERAL gets slim. GREETING gets none.
"""

from __future__ import annotations

import re
import sys
from datetime import datetime, timezone
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent))
from config import paths


def _load_text(path: Path, fallback: str = "") -> str:
    if path.exists():
        return path.read_text().strip()
    return fallback


def _channel_guidance(service: str) -> str:
    if service.upper() == "SMS":
        return (
            "── CHANNEL: SMS ──\n"
            "No read receipts or typing indicators. Keep messages under 320 characters\n"
            "(2 SMS segments) when possible. Use \"Reply YES to confirm\" for confirmations."
        )
    return ""


_RESPONSE_FORMAT = """\
── WHAT YOU CAN AND CANNOT DO ──
CAN: Generate SMS responses, apply family_file_updates (append/prepend/replace to sections that EXIST in the family file above — the system writes them immediately), flag needs_outreach (requests to text anyone whose phone number you know — team members or not), persist self_corrections to lessons.md (loaded into every future prompt).
CANNOT: Directly text people (outreach is sent shortly after this response, not in real-time — say "I'll message [name]" not "I'm texting them now"), access external systems, make medical decisions, see data outside your filtered context.
CRITICAL: Never claim you did something unless the family file above confirms it. If a section doesn't exist yet, you cannot update it — ask the coordinator to confirm the information and note that you'll save it.

── WHEN THINGS GO WRONG ──
If you previously sent an error message or glitch occurred, acknowledge it directly: "I hit a technical glitch — [what happened]. Here's what I was working on: [resume]." Never deflect or pretend it didn't happen. The coordinator can see everything.

── RESPONSE FORMAT ──
Respond with ONLY valid JSON matching the required schema. No markdown fencing, no explanation.

FIELD GUIDE:
- sms_response: The text message to send back. Keep under 320 chars when possible.
- internal_notes: Your reasoning (not shown to user).
- needs_outreach: Array of objects with phone (E.164 format: +1 then 10 digits, no dashes — e.g. +16514109390), name, message for people to contact. CRITICAL: If you say "I'll reach out" or "I'll message [name]" in sms_response, you MUST populate this array in the same response. If this array is empty, the outreach WILL NOT HAPPEN — there is no other mechanism. Say "I'll message [name]" in sms_response — never "I'm texting them now."
- family_file_updates: Array of objects with section, operation, content, old_content to update the family file. Operations: append, prepend, replace, resolve_issue. Only target sections that EXIST above.
- self_corrections: When the user corrects you, teaches you something, or says "remember that" / "don't do that again" / "that's wrong" — capture the lesson. The system writes these to lessons.md immediately; you will see them in your context on the next message. Prefix each with a category: [behavioral] how to reason/respond, [factual] care facts about this family, [operational] system behavior. Empty array if no correction this message.
- member_updates: Array of objects with section, operation, content, old_content to update the member's profile. Same format as family_file_updates. Use for personal preferences, communication style, etc. Empty array if nothing to update.
- routing_updates: Array of objects to register new family members. Only use when the COORDINATOR explicitly asks to add someone AND provides name + phone. Each object: action ("add"), phone (E.164), name, role (family_caregiver/professional_caregiver/community_supporter), relationship (to care recipient), access_level (full/limited). Empty array unless adding a member. REQUIRES coordinator confirmation before you populate this.

BEFORE YOU PROMISE TO CONTACT SOMEONE:
- Do you have their phone number? (check conversation history and family file)
- If yes: populate needs_outreach NOW. Don't just say you will — do it in this response.
- If no: tell the user you don't have the number and ask for it.
- Never say "I'll reach out" with an empty needs_outreach. That's a broken promise."""


_SECTION_RE = re.compile(r"^## ", re.MULTILINE)

# Sections to include per family context mode
_FAMILY_SECTIONS = {
    "family_full": None,  # None = include everything
    "family_meds": {
        "Active Medications", "Medication Hold Log", "Full Medication History",
        "Urgent Notes", "Care Recipient",
    },
    "family_team": {
        "Care Team", "This Week", "Urgent Notes", "Care Recipient",
    },
    "family_slim": "_slim",  # special: first 500 chars only
}

# Intent → which family context mode to use
_INTENT_FAMILY_MODE = {
    "EMERGENCY": "family_full",
    "ESCALATION": "family_full",
    "MEDICATION_CHANGE": "family_meds",
    "ONBOARDING": "family_team",
    "MULTI_MEMBER": "family_team",
    "GENERAL": "family_slim",
}


def _extract_family_sections(family_text: str, sections: set[str]) -> str:
    """Extract specific ## sections from family.md text by header name."""
    if not family_text or not sections:
        return ""

    parts = _SECTION_RE.split(family_text)
    # parts[0] is the preamble (before first ##), rest start with section name
    result = [parts[0].strip()] if parts[0].strip() else []

    for part in parts[1:]:
        lines = part.split("\n", 1)
        header = lines[0].strip()
        if header in sections:
            result.append(f"## {part.rstrip()}")

    return "\n\n".join(result)


def build_system_blocks(
    member: dict,
    family_context: str,
    service: str = "SMS",
    member_context: str = "",
    intent: str = "",
    tools_active: bool = False,
) -> list[dict]:
    """Build system prompt as ordered content blocks for cache-aware API calls.

    Returns list of dicts: {"type": "text", "text": str, "cache_breakpoint": bool}.
    Blocks 1-5 form the cached prefix (~4,987 tokens). Blocks 6-7 are dynamic.
    """
    blocks = []

    # Block 1: SOUL.md — agent identity (NEVER changes)
    soul_path = Path(__file__).parent.parent.parent / "SOUL.md"
    soul_text = _load_text(soul_path, "You are CareSupport — a care coordination agent.")
    blocks.append({"type": "text", "text": soul_text, "cache_breakpoint": False})

    # Block 2: Routing + Capabilities + Skills (rarely changes)
    parts = []
    agent_root = _load_text(paths.agent_root)
    if agent_root:
        parts.append(f"── ROUTING ──\n{agent_root}")

    capabilities = _load_text(paths.capabilities)
    if capabilities:
        parts.append(f"── CAPABILITIES ──\n{capabilities}")

    if paths.skills_dir.exists():
        skill_parts = []
        for skill_file in sorted(paths.skills_dir.glob("*.md")):
            skill_parts.append(skill_file.read_text().strip())
        if skill_parts:
            parts.append("── SKILLS ──\n" + "\n\n".join(skill_parts))

    if parts:
        blocks.append({"type": "text", "text": "\n\n".join(parts), "cache_breakpoint": False})

    # Block 3: Response format + channel guidance (NEVER changes)
    channel = _channel_guidance(service)
    block3_text = channel + "\n\n" + _RESPONSE_FORMAT if channel else _RESPONSE_FORMAT
    if tools_active:
        block3_text += (
            "\n\n── TOOLS ──\n"
            "You have tools to look up family information on demand. "
            "For greetings and simple conversation, respond directly without tools. "
            "For questions about schedule, medications, care team, or family notes, "
            "call the relevant tool first, then use the returned data in your response."
        )
    blocks.append({"type": "text", "text": block3_text, "cache_breakpoint": False})

    # Block 4: Lessons — global + family (changes weekly)
    lesson_parts = []
    family_dir = member.get("family_dir", "")
    if family_dir:
        family_lessons_path = Path(family_dir) / "lessons.md"
        if family_lessons_path.exists():
            entries = [l for l in family_lessons_path.read_text().split("\n") if l.startswith("- [")]
            if entries:
                lesson_parts.append("── FAMILY LESSONS (this family's corrections) ──\n" + "\n".join(entries))

    if paths.lessons.exists():
        entries = [l for l in paths.lessons.read_text().split("\n") if l.startswith("- [")]
        if entries:
            lesson_parts.append("── GLOBAL LESSONS (corrections from all conversations) ──\n" + "\n".join(entries))

    if lesson_parts:
        blocks.append({"type": "text", "text": "\n\n".join(lesson_parts), "cache_breakpoint": False})

    # Block 5: Member identity + member context (changes per member)
    # This is the LAST block in the cached prefix — cache_breakpoint=True
    member_lines = [
        f"YOU ARE TEXTING WITH: {member['name']} ({member['role']})",
        f"Their phone: {member['phone']}",
        f"Their access level: {member['access_level']}",
        f"Their relationship to care recipient: {member.get('relationship', member.get('role', 'unknown'))}",
    ]
    member_block = "\n".join(member_lines)
    if member_context:
        member_block += f"\n\n── WHAT YOU KNOW ABOUT {member['name'].upper()} ──\n{member_context}"

    blocks.append({"type": "text", "text": member_block, "cache_breakpoint": True})

    # Block 6: Family context — intent-driven loading
    family_mode = _INTENT_FAMILY_MODE.get(intent, "family_full")
    section_filter = _FAMILY_SECTIONS.get(family_mode)

    if family_context:
        if section_filter is None:
            # Full context (EMERGENCY, ESCALATION, or unknown intent)
            ctx = family_context
        elif section_filter == "_slim":
            # Slim: preamble only (name, overview, urgent notes snippet)
            ctx = family_context[:500].rsplit("\n", 1)[0]
        elif isinstance(section_filter, set):
            ctx = _extract_family_sections(family_context, section_filter)
        else:
            ctx = family_context

        if ctx.strip():
            blocks.append({
                "type": "text",
                "text": f"── FAMILY FILE (scoped to {member['name']}'s access level) ──\n{ctx}",
                "cache_breakpoint": False,
            })

    # Block 7: Current datetime (changes EVERY call — must be last)
    now = datetime.now(timezone.utc)
    blocks.append({
        "type": "text",
        "text": f"CURRENT DATE/TIME: {now.strftime('%A, %B %d, %Y at %I:%M %p')} CT",
        "cache_breakpoint": False,
    })

    return blocks


_LOG_LINE_RE = re.compile(
    r"^\[(\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2} UTC)\] \[(INBOUND|OUTBOUND)\] (.+)$",
    re.DOTALL,
)


def build_messages(user_message: str, conversation_history: str) -> list[dict]:
    """Parse flat conversation log into proper message turns, then append the new message.

    Log format: [2026-02-26 03:55:32 UTC] [INBOUND] Hello
    INBOUND → role: user, OUTBOUND → role: assistant.

    Multi-line messages (user sent a list of names, etc.) are joined back together.
    """
    messages: list[dict] = []

    if conversation_history and conversation_history != "[No conversation history]":
        # Reassemble multi-line log entries: lines not starting with [ belong to the previous entry
        raw_entries: list[str] = []
        for line in conversation_history.split("\n"):
            if line.startswith("[") and _LOG_LINE_RE.match(line):
                raw_entries.append(line)
            elif raw_entries:
                raw_entries[-1] += "\n" + line

        for entry in raw_entries:
            m = _LOG_LINE_RE.match(entry)
            if not m:
                continue
            direction = m.group(2)
            text = m.group(3).strip()
            if not text:
                continue
            role = "user" if direction == "INBOUND" else "assistant"

            # Collapse consecutive same-role messages
            if messages and messages[-1]["role"] == role:
                messages[-1]["content"] += "\n" + text
            else:
                messages.append({"role": role, "content": text})

    messages.append({"role": "user", "content": user_message})
    return messages


def system_blocks_to_string(blocks: list[dict]) -> str:
    """Flatten blocks back into a single string (for legacy/dry-run compatibility)."""
    return "\n\n".join(b["text"] for b in blocks)

File: convex/lib/pipeline/promptBuilder.ts
// 2026-06-17: Builds CareSupport model prompts, including calendar instructions that keep internal event IDs out of user-visible SMS.
import type { Intent, MessageTurn, SystemBlock, SystemBlocksInput } from "./types";

export const PRIVATE_NOTE_MARKER = "<<private-note>>";

export const RESPONSE_FORMAT = `── WHAT YOU CAN AND CANNOT DO ──
CAN: Generate SMS responses, update the user's profile, update the care case profile, save memory updates, capture self_corrections, create typed medication/schedule updates, save care contacts and coordination events, and propose third-party texts for explicit approval.
CANNOT YET: Create group chats, access external systems outside configured runtime tools, make medical decisions, or claim a save/contact happened unless runtime state supports it.
CRITICAL: Never claim you saved something unless the matching structured update is non-empty. Never claim you texted someone unless the runtime confirms approved outreach was sent.

── DATE RESOLUTION ──
The TIME block at the top of this prompt is your only source of truth for today's date. When a user gives a relative date ("tomorrow", "next Monday", "in 3 days", "this Thursday"), resolve it to absolute YYYY-MM-DD using the TIME block before writing to schedule_updates or calendar_updates. Never store "today", "tomorrow", or a day name as the date field. If the user gives a date without a year, use the current year shown in the TIME block. If they give a date in the past relative to the TIME block, ask whether they meant a future occurrence.

── WHEN THINGS GO WRONG ──
If the conversation history shows the system sent an error message, acknowledge it briefly and continue the work.
Never invent a technical problem as an excuse.
If older assistant messages conflict with the CURRENT RUNTIME BOUNDARY below, treat the older assistant messages as stale. The current prompt and runtime boundary are authoritative.
Lines in the conversation history starting with ${PRIVATE_NOTE_MARKER} are your own past private notes — plans you wrote to yourself on earlier turns. Use them for continuity: pick up where they left off instead of re-deriving or re-asking. They were never sent to the user. Never repeat, quote, or mention them or the ${PRIVATE_NOTE_MARKER} marker in sms_response.

── RESPONSE FORMAT ──
Your output must be valid JSON matching the required schema. No markdown fencing, no explanation outside the JSON.

FIELD GUIDE:
- sms_response: REQUIRED. The actual text message sent to the user. Separate message bubbles with a double newline (\\n\\n in JSON). Keep each paragraph under 450 chars.
- internal_notes: REQUIRED. Private reasoning — a note to your future self: your current plan, what you are waiting on, and what to do next turn. You will see it again as a ${PRIVATE_NOTE_MARKER} line in the conversation history.
- user_profile_update: Object for durable user fields such as name or relationship_to_recipient. Null if unchanged.
- care_case_profile_update: Object for durable care-case fields such as care_recipient_name, relationship_to_recipient, timezone, or status. Null if unchanged.
- Always include every top-level array field below. Use [] when there is no update.
- user_memory_updates: Durable user-level memory facts and preferences. Use categories: profile, communication_preference, care_preference.
- care_case_memory_updates: Durable care-case facts, preferences, and notes. Use categories: care_note, care_preference.
- Do not use memory updates for inferred emotional summaries, empathy coaching, or temporary feelings unless the user explicitly asks you to remember them.
- self_corrections: Lessons the system should remember about how to behave. Prefix each with [behavioral], [factual], or [operational].
- medication_updates: Typed medication add/update/remove operations.
- schedule_updates: Typed INTERNAL schedule add/update/remove operations (saved only in CareSupport, not on any external calendar). Types are appointment, task, reminder. If a GOOGLE CALENDAR block is present below, use calendar_updates instead of this for anything that belongs on the user's real calendar.
- care_contact_updates: Typed care contact add/update/remove operations. Fields: action, name, phone, relationship, contactType (family, professional_caregiver, agency, clinician, other), agencyName, role, availabilityNotes, contactPriority, canReceiveTexts, consentToContact, notes.
- coordination_event_updates: Typed coordination work add/update/remove operations. Fields: action, title, type (coverage_gap, schedule_change, handoff, task_followup, appointment, medication, outreach, other), status (open, waiting, resolved, cancelled), urgency, description, contactName, date, time.
- outreach_requests: Proposed third-party texts. Fields: contactName, purpose, message, coordinationEventTitle, approvalPrompt. This only creates a pending approval record; it does NOT send until the user later approves that exact message.
- calendar_updates: Writes to the user's REAL Google Calendar. Only available when a GOOGLE CALENDAR block appears below (calendar connected). Actions: create, update, delete. Fields: title, date (YYYY-MM-DD), startTime (HH:MM, 24h), endTime (HH:MM, 24h), location, eventId (required for update/delete), recurrence. On update, include eventId plus EVERY field you are changing (e.g. the new date and/or startTime) — fields you omit stay unchanged. recurrence is ONLY for repeating events ("every Friday", "daily"): set it to daily, weekdays, weekly, biweekly, monthly, or yearly; otherwise omit it entirely — never add recurrence just to move or retime an event. Updates/deletes apply to the whole recurring series.
- CALENDAR HONESTY: Never claim you added, moved, or removed anything on the user's Google Calendar unless a GOOGLE CALENDAR block appears below AND you returned a matching calendar_updates entry in this response. If there is no GOOGLE CALENDAR block, their calendar is not connected/available — do not say you put anything on it; offer to connect it (they can text "connect my calendar") or track it here instead.
- reactions: Optional tapbacks.
- effect: Optional iMessage effect.

CURRENT RUNTIME BOUNDARY:
- CareSupport is currently one trusted thread around one care situation.
- CareSupport can help text approved care contacts through separate one-to-one threads tied to this care case.
- If asked about pricing, say CareSupport is free during the concierge beta.
- If asked to add or contact another person, collect the missing name/phone/message details, save care-contact or coordination details through the available structured fields, and ask for explicit approval before outreach.
- If the person is not saved yet, say you can help add them first; do not say CareSupport cannot text them.
- Never claim a text was sent unless the runtime confirms approved outreach was sent.`;

export function channelGuidance(service: string): string {
  if (service.toUpperCase() === "SMS") {
    return [
      "── CHANNEL: iMessage/SMS ──",
      "They see read receipts and typing indicators.",
      "Plain text only — no markdown, no bullets, no headers.",
      "Most replies should fit in 1-2 message bubbles.",
    ].join("\n");
  }
  return "";
}

const LOG_LINE_RE = /^\[(\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2} UTC)\] \[(INBOUND|OUTBOUND)(?: (?:from|to) ([^\]]*))?\] ([\s\S]+?)$/;

export function buildMessages(userMessage: string, conversationHistory: string): MessageTurn[] {
  const messages: MessageTurn[] = [];

  if (conversationHistory && conversationHistory !== "[No conversation history]") {
    const rawEntries: string[] = [];
    for (const line of conversationHistory.split("\n")) {
      if (line.startsWith("[") && LOG_LINE_RE.test(line)) {
        rawEntries.push(line);
      } else if (rawEntries.length > 0) {
        rawEntries[rawEntries.length - 1] += `\n${line}`;
      }
    }

    for (const entry of rawEntries) {
      const match = LOG_LINE_RE.exec(entry);
      if (!match) continue;
      const direction = match[2];
      const attribution = match[3] ?? "";
      const text = match[4].trim();
      if (!text) continue;
      const role: "user" | "assistant" = direction === "INBOUND" ? "user" : "assistant";
      const labeled = role === "user" && attribution
        ? `[${attribution}]: ${text}`
        : stripStoredAssistantSpeakerPrefix(text);

      if (messages.length > 0 && messages[messages.length - 1].role === role) {
        messages[messages.length - 1].content += `\n${labeled}`;
      } else {
        messages.push({ role, content: labeled });
      }
    }
  }

  messages.push({ role: "user", content: userMessage });
  return messages;
}

function stripStoredAssistantSpeakerPrefix(text: string): string {
  return text.replace(/^\s*\[[^\]\n]{1,80}]:\s*/, "");
}

function intentGuidance(intent: Intent | string): string {
  switch (intent) {
    case "ONBOARDING":
      return [
        "── INTENT: ONBOARDING ──",
        "Ask one question at a time.",
        "Learn their name first, then who they are caring for, then the first thing to track.",
        "Write what you learn immediately through the structured update fields.",
      ].join("\n");
    case "MEDICATION_CHANGE":
      return [
        "── INTENT: MEDICATION ──",
        "Prefer medication_updates over generic care_case_memory_updates when the user gives specific medication instructions.",
      ].join("\n");
    case "BILLING":
      return [
        "── INTENT: BILLING ──",
        "Answer directly: CareSupport is currently free during the concierge beta.",
      ].join("\n");
    default:
      return "";
  }
}

export function buildSystemBlocks(input: SystemBlocksInput): SystemBlock[] {
  const blocks: SystemBlock[] = [];

  blocks.push({
    type: "text",
    text: input.soulContent || "You are CareSupport — a family care agent starting in one trusted text thread.",
    cacheBreakpoint: false,
  });

  if (input.isTestEnv) {
    blocks.push({
      type: "text",
      text: [
        "── ENVIRONMENT: TEST ──",
        "You are running in the TEST environment.",
        "Whenever you introduce or describe yourself to the user — a greeting, first contact, onboarding's first message, or answering who/what you are — begin that message with this exact line, on its own line:",
        "TEST ENVIRONMENT INITIALIZED",
        "Then a blank line, then your normal introduction.",
        "Only include this marker on self-introduction messages. Never add it to any other reply.",
      ].join("\n"),
      cacheBreakpoint: false,
    });
  }

  if (input.isFounder) {
    blocks.push({
      type: "text",
      text: [
        "── FOUNDER MODE ──",
        "You are talking with Liban — your founder and developer. You are software in active development, and with him you know it and say so plainly.",
        "When his message is feedback about CareSupport itself — tone, a bug, copy, behavior, a feature idea — respond as a collaborator who knows their own build, not as a care companion. Never perform onboarding or companion pleasantries at feedback.",
        "Capture every piece of such feedback in the dev_feedback field: an array of { category: \"tone\" | \"bug\" | \"feature\" | \"copy\", summary: <one sentence>, quote: <his words, verbatim or trimmed> }. The runtime files each item to the development queue (a GitHub issue) automatically.",
        "When you capture feedback, briefly confirm it in sms_response — e.g. \"Logged to the dev queue.\" — in your own voice, then answer whatever else his message needs.",
        "Receive statements as statements. When he tells you something about himself or about you, take it in and respond to it — do not pivot to \"What do you need?\" or re-ask what his message already answered. One insensitive re-ask cost real trust (issue #75).",
        "His real care-coordination requests are still real. Handle those exactly as you would for anyone.",
        "Only use dev_feedback in founder mode. It does not exist for anyone else.",
      ].join("\n"),
      cacheBreakpoint: false,
    });
  }

  const timeLines = [
    "── TIME ──",
    `Today is ${input.currentDateIso} (${input.currentDayOfWeek}).`,
    `Current time: ${input.currentTimeUtc} UTC.`,
    `Care recipient timezone: ${input.timezone || "UTC"}.`,
    "When the user says \"today\", \"tomorrow\", \"this week\", or names a weekday, resolve it against this anchor before writing any schedule_updates. Never store relative words as a date.",
  ];
  if (!input.timezoneConfirmed) {
    timeLines.push(
      "The timezone above is an unconfirmed default, not something the user told you. The FIRST time the user wants to schedule, add an appointment, or put anything on the calendar, do NOT create the event yet — first ask which city or area they're in (one short question). When they answer, set care_case_profile_update.timezone to the matching IANA timezone (e.g. Denver → America/Denver) and then carry out their scheduling request in that timezone. Ask this only once; never ask again after a timezone is set. For non-scheduling messages, do not bring up location.",
    );
  }
  blocks.push({
    type: "text",
    text: timeLines.join("\n"),
    cacheBreakpoint: false,
  });

  const operationalParts: string[] = [];
  if (input.routingContent) operationalParts.push(`── ROUTING ──\n${input.routingContent}`);
  if (input.capabilitiesContent) operationalParts.push(`── CAPABILITIES ──\n${input.capabilitiesContent}`);
  if (input.skillsContent) operationalParts.push(`── SKILLS ──\n${input.skillsContent}`);
  const intentBlock = intentGuidance(input.intent);
  if (intentBlock) operationalParts.push(intentBlock);
  if (operationalParts.length > 0) {
    blocks.push({
      type: "text",
      text: operationalParts.join("\n\n"),
      cacheBreakpoint: false,
    });
  }

  const channel = channelGuidance(input.service);
  blocks.push({
    type: "text",
    text: channel ? `${channel}\n\n${RESPONSE_FORMAT}` : RESPONSE_FORMAT,
    cacheBreakpoint: true,
  });

  if (input.isFounder) {
    blocks.push({
      type: "text",
      text: [
        "── FOUNDER-ONLY FIELD (extends the FIELD GUIDE above) ──",
        "- dev_feedback: Array of feedback items about CareSupport itself. Each: { category: \"tone\" | \"bug\" | \"feature\" | \"copy\", summary: <one sentence>, quote: <the founder's words> }. Include it whenever this message contains feedback about your behavior, tone, copy, or capabilities. Use [] when there is none.",
        "- TRUTHFULNESS: Never say feedback was logged, filed, or queued unless dev_feedback is non-empty in THIS response. The words do nothing — only the field files the issue.",
      ].join("\n"),
      cacheBreakpoint: false,
    });
  }

  if (input.lessonsContent) {
    blocks.push({ type: "text", text: input.lessonsContent, cacheBreakpoint: false });
  }

  const userLines = [
    `YOU ARE TEXTING WITH: ${input.user.name}`,
    `Phone: ${input.user.phone}`,
    `Relationship to care recipient: ${input.user.relationshipToRecipient ?? "unknown"}`,
    `User status: ${input.user.status}`,
  ];
  let userBlock = userLines.join("\n");
  if (input.userContext.trim()) {
    userBlock += `\n\n${input.userContext.trim()}`;
  }
  blocks.push({ type: "text", text: userBlock, cacheBreakpoint: true });

  if (input.careCaseContext.trim()) {
    blocks.push({
      type: "text",
      text: `── CARE CASE ──\n${input.careCaseContext.trim()}`,
      cacheBreakpoint: false,
    });
  }

  if (input.calendarContext) {
    blocks.push({
      type: "text",
      text: [
        "── GOOGLE CALENDAR (CONNECTED) ──",
        "Google Calendar is connected for this user. The events below are read live from their real Google Calendar — treat them as the source of truth when they ask what's on their calendar or schedule.",
        "",
        input.calendarContext,
        "",
        "WRITING TO THE CALENDAR: To add, move, retime, rename, or remove anything on the calendar you MUST return a calendar_updates entry — that is the ONLY thing that changes the real calendar. Your sms_response text changes nothing. Do not use schedule_updates for calendar items.",
        "CREATE: {\"action\":\"create\",\"title\":...,\"date\":\"YYYY-MM-DD\",\"startTime\":\"HH:MM\"} (24h, endTime optional).",
        "UPDATE (move/retime/rename): {\"action\":\"update\",\"eventId\":<id from list above>, plus the NEW values for whatever changed}. Moving to another day REQUIRES the new \"date\". Changing the time REQUIRES the new \"startTime\". An update with only action+eventId does NOTHING — always include the changed date/startTime. Example: move abc123 to June 9 → {\"action\":\"update\",\"eventId\":\"abc123\",\"date\":\"2026-06-09\"}.",
        "DELETE: {\"action\":\"delete\",\"eventId\":<id>} — confirm with the user first. Updates and deletes affect the whole recurring series.",
        "Event IDs are internal only. Use them only inside calendar_updates.eventId. Never include event IDs in sms_response when summarizing or listing calendar events.",
        "recurrence: set ONLY when the user wants the event to repeat (weekly, daily, etc.). For a one-time move/retime/rename, OMIT recurrence entirely — do not guess it.",
        "Only claim you changed the calendar if a matching calendar_updates entry with the changed fields is present in this response.",
      ].join("\n"),
      cacheBreakpoint: false,
    });
  }

  return blocks;
}

export function systemBlocksToString(blocks: SystemBlock[]): string {
  return blocks.map((block) => block.text).join("\n\n");
}
